Common/Securities/Security.cs

```diff
@@ -591,26 +591,25 @@ public void SetMarketPrice(BaseData data)
         /// on the data provided. Data is also stored into the security's data cache
         /// </summary>
         /// <param name="data">The security update data</param>
-        public void Update(IEnumerable<BaseData> data)
+        /// <param name="dataType">The data type</param>
+        /// <param name="containsFillForwardData">Flag indicating whether
+        /// <paramref name="data"/> contains any fill forward bar or not</param>
+        public void Update(IReadOnlyList<BaseData> data, Type dataType, bool? containsFillForwardData = null)
         {
-            BaseData last = null;
-            foreach (var grp in data.GroupBy(d => d.GetType()))
+            // use the last of each type to set market price
+            SetMarketPrice(data[data.Count - 1]);
+
+            var nonFillForwardData = data;
+            // maintaining regression requires us to NOT cache FF data
+            if (!containsFillForwardData.HasValue || containsFillForwardData.Value)
             {
-                var nonFillForward = new List<BaseData>();
-                foreach (var baseData in grp)
+                nonFillForwardData = data.Where(baseData => !baseData.IsFillForward).ToList();
```

Comment:
Surprised you want for the `Where` instead of `for`. Creating the list explicitly you could even set the size of it since we already know the max size, not sure if it would have an impact though.

Reply:
Yeah 👍  actually saw no difference after the `containsFillForwardData` flag was added, it really did the trick